src/eval.ml

```diff
@@ -165,6 +172,36 @@ and eq_modulo : term -> term -> bool = fun a b ->
   let res = try eq_modulo [(a,b)]; true with Exit -> false in
   if !log_enabled then log_eqmd (r_or_g res "%a == %a") pp a pp b; res
 
+(** [tree_walk t s] tries to match stack [s] against tree [t] *)
+and tree_walk : Dtree.t -> stack -> (term * stack) option = fun itree istk ->
+  let depth = Dtree.iter (fun _ -> 1) (fun _ chd ->
```

Comment:
Improve by pre-computing the depth and record its value in the Dtree structure itself.